Add module for deploying Play Framework 2.x projects

mattrco opened the pull request:
Adds the `play2` module for deploying projects using Play Framework 2.x

This functions in a similar fashion the the `play` module, however, there are some differences.
- `play dist` is invoked as a local build command. This produces a zip file in `dist/` with all project dependencies.
- This zip file is unzipped and rsynced to the project deployment directory

In order to only sync the files in `dist/{project-name}-{project-version}/`, an argument has been added to `rsync_from_local` to specify this.

Adding yet another argument to `rsync_from_local` (and thus to its caller, `fetch_render_copy`) is not particularly neat, so it would be good to start a discussion at some point on how we can tidy this module up.

There is also a fair bit of overlap between the `play` and `play2` functions for `supervisord` control, so it would be a good idea for these to be extracted into a separate generic module in future.

Commit:
Add play 2 root

Commit:
Add play2 option to env.lang check

Commit:
Add module for deploying play framework 2 projects

High level considerations:
* To deploy projects, a local build command `play dist` is invoked. This produces a zip file with all project dependencies
* This zip file is named `project-name`-`version` and appears in `dist/` the project directory.
* The play2 module will unzip this and rsync it to the application directory in deployment.

Commit:
Add local_path argument to rsync_from_local

This is required since `play dist` will create a zip file in `dist/` named `{project-name}-{version}.zip`

eazynow commented:
Looks good in theory.

I have no play2 projects at the moment. @MattCottingham, are you happy with all the functions working. I reckon it'd be good for someone else to try it first. @paologentili have you got one that you could test using this fork?

mattrco commented:
Sounds good - @paologentili I'll push the fabfile I have for our current project as soon as I've updated the docs.

Commit:
Add missing separator to local_build_path

This was causing rsync to sync the parent directory of lib/, causing classpath issues.

dcarley commented:
How do you feel about decoupling the `play dist` part from the default behaviour of `deploy_play2`?

At some point we'll want to fetch those artefacts from a repository in most cases. While still retaining the ability to deploy directly from local (ala. `dirty:true`).

mattrco commented:
It might be good to have both options, but the current puppet config restricts write access to the application's directory (apart from to `logs/`), so fetching artifacts after the release has been pushed would require changes to it.

`play dist` is also nice in that it resolves all dependencies before pushing the release - is there a case where we wouldn't want to do that?

dcarley commented:
Oh, I meant that we would still use the ZIP created by `play dist`, but we wouldn't necessarily want to create the ZIP at the point of deployment. It would normally be created prior by CI and stored in something like Artifactory.

mattrco commented:
Ah, I understand. Sure, I can make an optional artifactory path, so that if it isn't provided it will attempt to run `play dist` in the project directory.

mattrco commented:
It looks like we already have some code for dealing with artifactory resources at https://github.com/YellLabs/yellfabric/blob/master/operations.py#L84

I'd prefer to adapt this and reuse it in the Play 2 module since it does most of what we want, but we would need to take into account all the other projects using it.

For that reason it might be easier to merge this PR and open a separate issue for optional artifactory resources.

dcarley commented:
We don't need the Artifactory functionality right now. I was just keen to have the `play dist` step broken out to a separate command now, so that we don't depend on it.

Commit:
Extract package and unzip actions into separate functions

Decoupling these actions will make it easier to use an existing zip package for deployment instead of running `play dist` first.

Commit:
Use `dist` switch to determine whether to execute `play dist` in build command

Build command will execute `play dist` if dist=True, otherwise it will only
attempt to unzip the application zip.

mattrco commented:
I've extracted the `play dist` and `unzip` commands into separate functions. The default behaviour (only unzip and push package) can be modified with the `dist` boolean on `deploy_play2`.

The only thing that remains is making the zipfile path configurable, though we're coupled to how `play dist` organises the archive so it may be best to defer this to when we build in full artifactory support.

Commit:
Remove unused imports and fix line length
